packages/assets-controllers/src/AssetsContractController.ts

```diff
@@ -62,137 +67,257 @@ export const SINGLE_CALL_BALANCES_ADDRESS_BY_CHAINID: Record<Hex, string> = {
-      const selectedNetworkClient = getNetworkClientById(
-        selectedNetworkClientId,
-      );
-      const { chainId } = selectedNetworkClient.configuration;
+  constructor({
+    messenger,
+    chainId: initialChainId,
+  }: {
+    messenger: AssetsContractControllerMessenger;
+    chainId: Hex;
+  }) {
+    this.messagingSystem = messenger;
+    this.#provider = undefined;
+    this.#ipfsGateway = IPFS_DEFAULT_GATEWAY_URL;
+    this.#chainId = initialChainId;
+
+    this.#registerActionHandlers();
+    this.#registerEventSubscriptions();
+  }
 
-      if (this.config.chainId !== chainId) {
-        this.configure({
-          chainId: selectedNetworkClient.configuration.chainId,
-        });
-      }
-    });
+  // TODO: Expand into base-controller utility function that batch registers action handlers.
+  #registerActionHandlers() {
+    const nonMethodClassProperties = [
+      'constructor',
+      'messagingSystem',
+      'setProvider',
+      'provider',
+      'ipfsGateway',
+      'chainId',
+    ];
+
+    getKnownPropertyNames<keyof this>(Object.getPrototypeOf(this)).forEach(
+      (method) => {
+        if (
+          ((key: keyof this): key is AssetsContractControllerMethodName =>
+            !nonMethodClassProperties.find((e) => e === key) &&
+            typeof this[key] === 'function')(method)
+        ) {
+          this.messagingSystem.registerActionHandler(
+            `${name}:${method}`,
+            // TODO: Write a generic for-loop implementation that iterates over an input union type in tandem with the input array.
+            // @ts-expect-error Both assigned argument and assignee parameter are using the entire union type for `method` instead of the type for the current element
+            this[method].bind(this),
+          );
+        }
+      },
+    );
+  }
+
+  #registerEventSubscriptions() {
+    this.messagingSystem.subscribe(
+      `PreferencesController:stateChange`,
+      ({ ipfsGateway }) => {
+        this.#ipfsGateway = ipfsGateway;
+      },
+    );
+
+    this.messagingSystem.subscribe(
+      `NetworkController:networkDidChange`,
+      ({ selectedNetworkClientId }) => {
+        const chainId = this.#getCorrectChainId(selectedNetworkClientId);
+
+        if (this.#chainId !== chainId) {
+          this.#chainId = chainId;
+          // @ts-expect-error TODO: remove this annotation once the `Eip1193Provider` class is released
+          this.#provider = this.#getCorrectProvider();
+        }
+      },
+    );
   }
 
   /**
    * Sets a new provider.
    *
-   * TODO: Replace this wth a method.
-   *
-   * @property provider - Provider used to create a new underlying Web3 instance
+   * @param provider - Provider used to create a new underlying Web3 instance
    */
-  set provider(provider: Provider) {
-    this._provider = provider;
+  setProvider(provider: Provider | undefined) {
+    this.#provider = provider;
   }
 
   get provider() {
```

Comment:
Yes I left it because of a test checking for the error being thrown, but makes sense to get rid of that as well.

Removed: 418df796688bf6da7cf8946489e90f30318d648f